app/services/definitions.py

```diff
@@ -0,0 +1,543 @@
+
+    bound_logger.debug("Getting schemas associated with fetched ids")
+    schemas = await schema_futures(schema_ids, aries_controller)
+
+    if schema_issuer_did:
+        schemas = [
+            schema for schema in schemas if schema.id.split(":")[0] == schema_issuer_did
+        ]
+    if schema_name:
+        schemas = [schema for schema in schemas if schema.name == schema_name]
+    if schema_version:
+        schemas = [schema for schema in schemas if schema.version == schema_version]
+
+    return schemas
+
+
+async def get_schemas_governance(
+    aries_controller: AcaPyClient,
+    schema_id: Optional[str],
+    schema_issuer_did: Optional[str],
+    schema_name: Optional[str],
+    schema_version: Optional[str],
+) -> List[CredentialSchema]:
+    """
+    Governance agents gets all schemas created by itself
+    """
+    bound_logger = logger.bind(
+        body={
+            "schema_id": schema_id,
+            "schema_issuer_did": schema_issuer_did,
+            "schema_name": schema_name,
+            "schema_version": schema_version,
+        }
+    )
+    # Get all created schema ids that match the filter
+    bound_logger.debug("Fetching created schemas")
+    response = await handle_acapy_call(
+        logger=bound_logger,
+        acapy_call=aries_controller.schema.get_created_schemas,
+        schema_id=schema_id,
+        schema_issuer_did=schema_issuer_did,
+        schema_name=schema_name,
+        schema_version=schema_version,
+    )
+
+    # Initiate retrieving all schemas
+    schema_ids = response.schema_ids or []
+
+    bound_logger.debug("Getting schemas associated with fetched ids")
+    schemas = await schema_futures(schema_ids, aries_controller)
+
+    return schemas
+
+
+async def schema_futures(
+    schema_ids: List[str], aries_controller: AcaPyClient
+) -> List[CredentialSchema]:
+    """
+    Get schemas with attributes from schema ids
+    """
+    logger.debug("Fetching schemas from schema ids")
+    # We now have schema_ids; the following logic is the same whether called by governance or tenant.
+    # Now fetch relevant schemas from ledger:
+    get_schema_futures = [
+        handle_acapy_call(
+            logger=logger,
+            acapy_call=aries_controller.schema.get_schema,
+            schema_id=schema_id,
+        )
+        for schema_id in schema_ids
+    ]
+
+    # Wait for completion of retrieval and transform all schemas into response model (if a schema was returned)
+    if get_schema_futures:
+        logger.debug("Fetching each of the created schemas")
+        schema_results: List[SchemaGetResult] = await asyncio.gather(
+            *get_schema_futures
+        )
+    else:
+        logger.debug("No created schema ids returned")
+        schema_results = []
+
+    schemas = [
+        credential_schema_from_acapy(schema.var_schema)
+        for schema in schema_results
+        if schema.var_schema
+    ]
+
+    return schemas
+
+
+class CredDefPublisher:
```

Comment:
See this class makes sense -- I just think maybe `check_endorser_connection` can be moved to a util method as well, because there may be more contexts that also need to assert a controller has a connection with the endorser -- so it can be re-used, not just in creating cred defs. Same with wait for transaction ack, I can imagine that being reused elsewhere too. But wait_for_rev_reg, perfect, makes sense for that to be part of a CredDefPublisher. I would just move it to its own module, and perhaps rename to CredentialDefinitionPublisher. Long name, but the keystrokes are free, and I think verbosity (within reason) can help with maintainability somewhat